src/benchmarklib/tpcc/helper.cpp

```diff
@@ -10,7 +10,7 @@
 namespace tpcc {
 
 void execute_tasks_with_context(std::vector<std::shared_ptr<opossum::OperatorTask>>& tasks,
-                                std::shared_ptr<opossum::TransactionContext> t_context) {
+                                const std::shared_ptr<opossum::TransactionContext>& t_context) {
```

Comment:
It is not `execute_tasks_with_context` that takes ownership, but someone down the line.

Reply:
Okay but from the view point of the user, it doesn’t matter if `execute_tasks_with_context` takes ownership of it or some method down the call tree and `execute_tasks_with_context ` itself knows that it is going to call another method that takes ownership.

Reply:
No it doesn’t. I was thinking the same. What however would make a difference is, if we passed by const& when we didn’t need ownership. This way, the caller would have to dereference the pointer so it would already be clear from the calling code (`method(*object_ptr)`).

Reply:
This is the discussion that we had before, right? It works if we know that we have a value. If the pointer could be null, we would need to pass a raw pointer and we want to avoid those. C++ should have a "somewhat" raw pointer that is like a nullable reference, but prohibits you from using pointer magic.

I'll take the "take it by copy if you want to own it" out of the flowchart. The less rules, the better.

Reply:
Does this make more sense?

![flowchart_arguments](https://user-images.githubusercontent.com/575106/33320811-b506fa16-d443-11e7-9aba-0879fd070b31.png)

Reply:
I find this still a bit confusing. The 

What’s more, the part under “Is it a smart pointer – No” isn’t correct. Moving a const& variable into something does exactly nothing, because in order to move, the variable needs to be non-const. Moving a & variable is dangerous, because you invalidate a variable without the awareness of the caller and he will probably continue using that variable afterwards.

I’m not sure why we even need our own guidelines for simple (non-pointer) data types. The early mentioned guidelines explain very well how to pass arguments. (https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#fcall-parameter-passing) This part has very little to do with pointers, they explain simple default argument passing as well as advanced stuff if needed, the guidelines have 15.000+ stars and they come directly from Bjarne Stroustrup, the inventor of C++. I’d rather point to these guidelines and then add our project specific things like avoid raw-pointers.

Reply:
I'll need to read this carefully again and think about this for a while.